### ISSUE TITLE
Filter card summary design

### ISSUE THREAD

**Issue by asbates** [insightsengineering/teal.slice#220]
Review and improve the design of the filter card summary.

Some comments from @nikolas-burkoff 

Arrow would be better for displaying selected range of numeric columns.

![image](https://user-images.githubusercontent.com/36482959/222779533-5413ec6e-e734-4f38-98e3-96411893193c.png)

If there aren't any NA/Inf maybe we shouldn't show NA/Inf

![image](https://user-images.githubusercontent.com/36482959/222779605-ffbdf370-7b19-40c3-bc92-1a32e1c2430d.png)

**Comment by lcd2yyz** [comment]
Also would like to add some conditional handling in the summary for categorical variables (checkbox groups UI)
- If all available levels are selected (basically no filtering applied), then display "All levels selected". 
- If only 1-2 levels are selected (for variable with more than 2 levels available), display the full selected value (eg. `M, F` for variable SEX)
- If >=3 or more levels selected, then keep display as-is

![image](https://user-images.githubusercontent.com/78242757/224216741-cdd8a307-103b-49d2-acca-ff239dbbcac2.png)

**Comment by asbates** [comment]
For NA/Inf, I don't think displaying just "NA" or "Inf" conveys much information. What I mean is I don't think "NA" conveys to the user that missing values are kept. We should rethink not just the alignment here but how the message is conveyed.

**Comment by lcd2yyz** [comment]
Yes agree, I thought it was something still WIP, so I didn't comment. 
So can we go back to this? Originally posted to #181 
![image](https://user-images.githubusercontent.com/78242757/225787479-6d1f669b-9775-4216-afc9-93bfde93ef26.png)

We can right align the text to show `Includes: NA, Inf` with a check or cross mark behind the NA/Inf. If there are no NA/Inf values in the data, then don't show in "Includes"

**Comment by donyunardi** [comment]
###  Levels selected
> Also would like to add some conditional handling in the summary for categorical variables (checkbox groups UI)
> 
> * If all available levels are selected (basically no filtering applied), then display "All levels selected".
> * If only 1-2 levels are selected (for variable with more than 2 levels available), display the full selected value (eg. `M, F` for variable SEX)
> * If >=3 or more levels selected, then keep display as-is

I think we talked about this issue before, and it has to do with variables that have really long values. For example, the RACE variable has options like "American Indian or Alaska Native" and "Native Hawaiian or Other Pacific Islander." If someone picks both of these options, it might be hard to fit them both on one line when we show them on a card.

One idea we had was to cut off the variable at a certain length. But we realized that this might not be the best solution if the first option is really long, because then the second option might not show up.

### NA / Inf

> For NA/Inf, I don't think displaying just "NA" or "Inf" conveys much information. What I mean is I don't think "NA" conveys to the user that missing values are kept. We should rethink not just the alignment here but how the message is conveyed.

> Yes agree, I thought it was something still WIP, so I didn't comment. So can we go back to this? Originally posted to #181 
> We can right align the text to show `Includes: NA, Inf` with a check or cross mark behind the NA/Inf. If there are no NA/Inf values in the data, then don't show in "Includes"

Couple thoughts on this:
1. I'm trying to determine whether adding the word "Include" would enhance clarity. At present, we're employing the term "Keep" in the selection. Should we keep it the same way?
    <img src="https://user-images.githubusercontent.com/8597300/227066734-4e0a3550-bd70-49b8-a118-82b4841563b0.png" width="50%" height="50%" />

2. Currently, we display both the "from" and "to" values for dates. However, having an "NA" on the right side does not appear visually appealing to me.
    <img src="https://user-images.githubusercontent.com/8597300/227067317-4ba74d30-7dbb-4a42-ae80-f9b7df1e504b.png" width="50%" height="50%" />
3. Should we consider providing a sentence instead? "Data includes NA and infinite values"
4. Should the information about NA and Inf added on the new line after "x levels selected" label?

**Comment by donyunardi** [comment]
### Acceptance Criteria
* Finalize design of filter card summary
* List out all task needed to make the update.

This is a research ticket and we will execute on different issue.

**Comment by lcd2yyz** [comment]
I will try to draft up some more harmonized and standardized filter card designs. 
Please help to identify all the possible combinations of filter class/type/state that we may have. Following are the list of the top of my head.

By filter class
- [ ] Interactive (with default set by API)
- [ ] Interactive (user created in-app)
- [ ] Fixed 
- [ ] Grouped?

By filter variable type
- [ ] Numeric
- [ ] Dates/datetime
- [ ] Factors
- [ ] Characters (eg. more than 5 factors?)

By filter card state
- [ ] Card expanded
- [ ] Card collapsed

Any more I missed?

**Comment by donyunardi** [comment]
I think you covered most of them. The only one that I think of is `logical` variable type.

**Comment by chlebowa** [comment]
> By filter class
> 
> * [ ]  Interactive (with default set by API)
> * [ ]  Interactive (user created in-app)
> * [ ]  Fixed
> * [ ]  Grouped?

This no longer reflects our design. There are no separate classes for interactive/fixed filter states and no separate classes for filters created with API on start-up or by the user during runtime.

There is `FilterState` class (with subclasses for different data types).
A `FilterState` object defines subsetting on one variable only. 
A `FilterState` can be created on start-up, by the API (as defined by the app developer) or during runtime, by the user.
A `FilterState` can be disabled and re-enabled by the user.
A disabled `FilterState` cannot be modified.
A `FilterState` created by the API can be created disabled. This can be changed by the user.
A `FilterState` created by the API can be created fixed, which means none of its properties can be modified. It can still be disabled.

> By filter variable type
> 
> * [ ]  Numeric
> * [ ]  Dates/datetime
> * [ ]  Logical
> * [ ]  Factors
> * [ ]  Characters (eg. more than 5 factors?)

All variable types except `logical` create a `ChoicesFilterState` if there are less than N unique values, N being defined by a `teal` option nad being constant throughout the session. `ChoicesFilterStates` always converts data to a factor allows choosing values with checkboxes*. When >N unique values are present in the data, selection defines ranges, whether the data is numeric, date, or datetime.
(* If a factor is passed that has >N unique values, a drop-down is used instead.)


> By filter card state
> 
> * [ ]  Card expanded
> * [ ]  Card collapsed
> 
> Any more I missed?

**Comment by lcd2yyz** [comment]
Finalized the filter card summary design and shared general thoughts with @donyunardi

## First, for all interactive filters - by variable type
![image](https://user-images.githubusercontent.com/78242757/232963702-67045425-d65d-4b15-a959-1b02713a39c9.png)
![image](https://user-images.githubusercontent.com/78242757/232963765-894a48c2-24f8-4956-80d2-3e16a331e88e.png)

## For fixed filters, with single variable in condition
![image](https://user-images.githubusercontent.com/78242757/232963811-c5c6795b-910a-4d4f-a902-a62d9ae33199.png)

## For grouped filters
![image](https://user-images.githubusercontent.com/78242757/232963872-95e76459-15ff-423c-8b27-e9de5b6ee522.png)
![image](https://user-images.githubusercontent.com/78242757/232963899-27a5461c-8dc3-49d7-99e8-f12ae11a7b97.png)

## Lastly for a case that generated some debates - fixed filters with multi-variable conditions
![image](https://user-images.githubusercontent.com/78242757/232964067-9234ac72-c58c-4611-9399-ee0b34f9089a.png)
![image](https://user-images.githubusercontent.com/78242757/232964161-ee515d1d-c8c3-4d1b-a7da-32f2d5a16e43.png)
![image](https://user-images.githubusercontent.com/78242757/232964183-1b767e28-9df8-410c-9c9d-78b69b3fe5c0.png)

Sorry for all the screenshots, I did this in PPT and not able to share an internal doc here. I will send gSlide link via group chat.

**Comment by chlebowa** [comment]
Currently the summary of a disabled filter card spells "Disabled". Are we to depart from that? Please confirm.

**Comment by lcd2yyz** [comment]
> Currently the summary of a disabled filter card spells "Disabled". Are we to depart from that? Please confirm.

Thank you very much for pointing it out! Let's go with active/inactive when we talk about state of the filter cards. Modified in the mockup above as well.

**Comment by lcd2yyz** [comment]
> 4\. I think for now we should focus on content.

Agree focus should be on content and content arrangement as priority. Coloring per bootstrap theming is nice to have at this point, perhaps can be tackled together with the theming of the UI component (eg. progress bar/distribution) later. Will open a separate issue.

**Comment by asbates** [comment]
I can't remember if this was discussed but I think displaying "NA" in the selected value section is redundant.

![image](https://user-images.githubusercontent.com/36482959/234987568-2ab683c6-3f41-473e-9c07-c4e27904e774.png)


I'd recommend sticking with just the one 'NA' on the right side, like so.

![image](https://user-images.githubusercontent.com/36482959/234987784-99a5dcd8-f597-453c-8d6d-0443e918c359.png)


@lcd2yyz is that OK with you?

**Comment by lcd2yyz** [comment]
@asbates Yes agree - should display NA right-aligned only. Thanks!

**Comment by lcd2yyz** [comment]
@asbates Sorry one more minor detail... if the NA is an explicit data point as a character string `"NA"`, not the missing data point `NA` or `NA_character`, then it should still show up in the selected value section (in addition to the right-aligned NA-checkmark if there is missing data). Hope that make sense?

**Comment by chlebowa** [comment]
That would be the intended behavior for a character/factor.

**Comment by asbates** [comment]
@lcd2yyz can you provide some examples of input and output regarding numeric values and scientific notation?

**Comment by lcd2yyz** [comment]
> @lcd2yyz can you provide some examples of input and output regarding numeric values and scientific notation?

Added an example in this issue https://github.com/insightsengineering/teal.slice/issues/251. We can continue discussion about numeric value and scientific notation there?
